### PR TITLE
Covariate analysis

### PR DESCRIPTION
**Pull request by PierreBoyeau**

### CODE REVIEW

File: src/scvi_v2/_module.py [hunk 1]
```diff
@@ -27,6 +27,12 @@
 _normal_initializer = jax.nn.initializers.normal(stddev=0.1)
 
 
+@jax.jit
+def exp(x: NdArray) -> NdArray:
+    """Exponential function."""
+    return jnp.exp(x)
```
Comment (adamgayoso) [inline src/scvi_v2/_module.py]: what is this for?

File: src/scvi_v2/_module.py [hunk 2]
```diff
@@ -97,35 +99,51 @@ def __call__(
 
         has_mc_samples = z.ndim == 3
         z_stop = z if not self.stop_gradients else jax.lax.stop_gradient(z)
-        z_ = nn.LayerNorm(name="u_ln")(z_stop)
+
+        if self.use_conditional_norm:
+            z_dim = z.shape[-1]
+            z_ = ConditionalNormalization(z_dim, self.n_batch)(z_stop, batch_covariate, training=training)
```
Comment (justjhong) [inline src/scvi_v2/_module.py]: if we do CBN here, and the batch delta is a function of z, wouldn't this reduce the effectiveness/importance of the batch deltas?

File: src/scvi_v2/_module.py [hunk 3]
```diff
@@ -262,7 +288,11 @@ def __call__(self, u: NdArray, sample_covariate: NdArray, training: Optional[boo
         sample_embed = nn.Embed(self.n_sample, self.n_latent_sample, embedding_init=_normal_initializer)(
             sample_covariate
         )  # (batch, n_latent_sample)
-        sample_embed = nn.LayerNorm(name="sample_embed_ln")(sample_embed)
+        if self.use_conditional_norm:
+            sample_embed = ConditionalNormalization(self.n_latent_sample, self.n_batch)(
```
Comment (justjhong) [inline src/scvi_v2/_module.py]: this seems more reasonable, but as we discussed in our meeting we may not need it at all

File: src/scvi_v2/_model.py [hunk 4]
```diff
@@ -610,6 +614,219 @@ def get_local_sample_distances(
...
+        store_lfc
+            Whether to store the log-fold changes in the module.
+            Storing log-fold changes is memory-intensive and may require to specify
+            a smaller set of cells to analyze, e.g., by specifying `adata`.
+        """
+        adata = self.adata if adata is None else adata
+        self._check_if_trained(warn=False)
+        # Hack to ensure new AnnDatas have indices.
+        if "_indices" not in adata.obs:
+            adata.obs["_indices"] = np.arange(adata.n_obs).astype(int)
```
Comment (justjhong) [inline src/scvi_v2/_model.py:831]: maybe we can stick this in validate anndata

File: src/scvi_v2/_model.py [hunk 5]
```diff
@@ -610,6 +614,219 @@ def get_local_sample_distances(
...
+
+                def per_sample_inference_fn(pair):
+                    rngs, cf_sample = pair
+                    return inference_fn(rngs, cf_sample)
+
+                eps_ = jax.lax.transpose(jax.lax.map(per_sample_inference_fn, (stacked_rngs, cf_sample)), (1, 0, 2))
+            eps_ = eps_[:, donor_mask]
+            eps = (eps_ - eps_.mean(axis=1, keepdims=True)) / (1e-6 + eps_.std(axis=1, keepdims=True))  # over samples
+
+            betas = jnp.einsum("ks,nsd->nkd", Amat, eps)
```
Comment (justjhong) [inline src/scvi_v2/_model.py]: add Amat and prefactor as static args

File: src/scvi_v2/_model.py [hunk 6]
```diff
@@ -610,6 +614,219 @@ def get_local_sample_distances(
...
+            # Statistical test
+            betas_norm = jnp.einsum("nkd,kl->nld", betas, prefactor)
+            df = betas_norm.shape[-1]
+            ts = (betas_norm**2).sum(axis=-1)
+            pvals = 1 - jax.scipy.stats.chi2.cdf(ts, df=df)
+
+            # Decoding
+            betas_ = betas.transpose((1, 0, 2))  # (n_metadata, n_cells, n_latent)
+            betas_ = betas_ * eps_.std(axis=1)
+            lfcs = jnp.einsum("lg,knl->kng", self.module.params["px"]["Dense_0"]["kernel"], betas_)
```
Comment (justjhong) [inline src/scvi_v2/_model.py]: you can only do this for low dim batch right? if so, should enforce that

File: src/scvi_v2/_model.py [hunk 7]
```diff
@@ -610,6 +614,219 @@ def get_local_sample_distances(
...
+        Xmat_names = np.concatenate(Xmat_names)
+        Xmat = np.concatenate(Xmat, axis=1).astype(np.float32)
+        Xmat = Xmat[donor_mask]
+
+        if normalize_design_matrix:
+            Xmat = (Xmat - Xmat.mean(axis=0)) / (1e-6 + Xmat.std(axis=0))
+        if offset_design_matrix:
+            Xmat = np.concatenate([np.ones((Xmat.shape[0], 1)), Xmat], axis=1)
+            Xmat_names = np.concatenate([np.array(["offset"]), Xmat_names])
+        prefactor = sqrtm(Xmat.T @ Xmat)
```
Comment (justjhong) [inline src/scvi_v2/_model.py]: where does the prefactor come from?

**Review by justjhong — COMMENTED** [review]
Looks great! have to say the perform_multivariate_analysis fcn is a monster though. if we go through w it we will have to refactor it :)

File: src/scvi_v2/_module.py (hunk 3 above)
Comment (PierreBoyeau) [reply src/scvi_v2/_module.py]: I agree, all these new conditional normalization correspond to stuff I tried and forgot to remove. Should I remove them all?

File: src/scvi_v2/_model.py (hunk 7 above)
Comment (PierreBoyeau) [reply src/scvi_v2/_model.py]: See Equation (7) (the one giving the asymptotic distribution of beta) from the overleaf note I have, let me send it to you over slack

File: src/scvi_v2/_model.py (hunk 6 above)
Comment (PierreBoyeau) [reply src/scvi_v2/_model.py]: Yes! The LFC estimation is trash I think. @cane11 is working on a fix for this on branch `can_multires`, but from what he has told me, this needs some cleanup.

Comment (PierreBoyeau) [reply src/scvi_v2/_model.py]: basically, the right approach seems to be to directly compute the LFC, by decoding some base $$z$$ and $$z + \beta$$, and computing the associated LFC.

Comment (PierreBoyeau) [reply src/scvi_v2/_model.py]: With this fix, this approach should work with any decoder setup.

Comment (justjhong) [reply src/scvi_v2/_model.py]: yes i like this better!

File: src/scvi_v2/_module.py (hunk 3 above)
Comment (justjhong) [reply src/scvi_v2/_module.py]: yes i think best to remove before merging

File: src/scvi_v2/_module.py [hunk 8]
```diff
@@ -558,3 +572,18 @@ def compute_h_from_x(self, x, sample_index, batch_index, cf_sample=None, continu
         }
         generative_outputs = self.generative(**generative_inputs)
         return generative_outputs["h"]
+
+    def compute_h_from_x_eps(
+        self, x, sample_index, batch_index, extra_eps, cf_sample=None, continuous_covs=None, mc_samples=10
+    ):
+        """Compute normalized gene expression from observations using predefined eps"""
```
Comment (justjhong) [inline src/scvi_v2/_module.py:574]: add comment on what extra_eps is and what its used for. maybe this should default to None and if none we use "z" instead for general use?

File: src/scvi_v2/_module.py [hunk 9]
```diff
@@ -94,37 +95,49 @@ def __call__(
...
+            )  # (batch, n_latent_sample)
+            batch_embed = nn.LayerNorm(name="batch_embed_ln")(batch_embed)
+            if has_mc_samples:
+                batch_embed = jnp.tile(batch_embed, (z_.shape[0], 1, 1))
 
-        if self.low_dim_batch:
-            mu = nn.Dense(self.n_out)(z + residual)
+            res_dim = self.n_in if self.low_dim_batch else self.n_out
+
+            if self.batch_values:
```
Comment (justjhong) [inline src/scvi_v2/_module.py]: this isnt super relevant to the covariate analysis right?

File: src/scvi_v2/_model.py [hunk 10]
```diff
@@ -641,13 +646,413 @@ def get_local_sample_distances(
...
-            reductions,
-            adata=adata,
-            batch_size=batch_size,
-            use_vmap=use_vmap,
-            norm=norm,
-            mc_samples=mc_samples,
+            reductions, adata=adata, batch_size=batch_size, use_vmap=use_vmap, norm=norm, mc_samples=mc_samples
+        )
+
+    def get_aggregated_posterior(
```
Comment (justjhong) [inline src/scvi_v2/_model.py:650]: review this when you get a chance (my code) @PierreBoyeau !

File: src/scvi_v2/_model.py [hunk 11]
```diff
@@ -641,13 +646,413 @@ def get_local_sample_distances(
...
+            }
+        )
+        return results
+
+    def perform_multivariate_analysis(
+        self,
+        adata=None,
+        donor_keys: List[Tuple] = None,
+        donor_subset: Optional[List[str]] = None,
+        batch_size=256,
```
Comment (justjhong) [inline src/scvi_v2/_model.py]: nit: batch_size: int = 256

File: src/scvi_v2/_model.py [hunk 12]
```diff
@@ -641,13 +646,413 @@ def get_local_sample_distances(
...
+                "log_probs": log_probs,
+                "log_ratios": log_ratios,
+                "is_admissible": cell_sample_admissible,
+            }
+        )
+        return results
+
+    def perform_multivariate_analysis(
+        self,
+        adata=None,
```
Comment (justjhong) [inline src/scvi_v2/_model.py]: nit: add typing

File: src/scvi_v2/_model.py [hunk 13]
```diff
@@ -641,13 +646,413 @@ def get_local_sample_distances(
...
+            Whether to store the expression in the module if logfoldchanges are computed.
+        """
+        adata = self.adata if adata is None else adata
+        self._check_if_trained(warn=False)
+        # Hack to ensure new AnnDatas have indices.
+        if "_indices" not in adata.obs:
+            adata.obs["_indices"] = np.arange(adata.n_obs).astype(int)
+
+        adata = self._validate_anndata(adata)
+        scdl = self._make_data_loader(adata=adata, indices=None, batch_size=batch_size, iter_ndarray=True)
```
Comment (justjhong) [inline src/scvi_v2/_model.py:834]: keep indices as a parameter?

File: src/scvi_v2/_model.py [hunk 14]
```diff
@@ -641,13 +646,413 @@ def get_local_sample_distances(
...
+        n_sample = self.summary_stats.n_sample
+        vars_in = {"params": self.module.params, **self.module.state}
+
+        donor_mask = (
+            np.isin(self.sample_order, donor_subset) if donor_subset is not None else np.ones(n_sample, dtype=bool)
         )
+        donor_mask = np.array(donor_mask)
+
+        @partial(jax.jit, static_argnames=["use_mean", "Amat", "prefactor"])
```
Comment (justjhong) [inline src/scvi_v2/_model.py]: need to add Amat and prefactor as args here

File: src/scvi_v2/_model.py [hunk 15]
```diff
@@ -641,13 +646,413 @@ def get_local_sample_distances(
...
+                stacked_rngs=stacked_rngs,
+                x=jnp.array(inf_inputs["x"]),
+                sample_index=jnp.array(inf_inputs["sample_index"]),
+                batch_index=jnp.array(array_dict[REGISTRY_KEYS.BATCH_KEY]),
+                continuous_covs=continuous_covs,
+                cf_sample=jnp.array(cf_sample),
+                use_mean=False,
+                stacked_rngs_de=stacked_rngs_de,
+            )  # (n_cells, n_donors, n_latent)
+            beta.append(np.array(res["beta"]))
```
Comment (justjhong) [inline src/scvi_v2/_model.py:1013]: im surprised you didnt have to do jax.device_get here

File: src/scvi_v2/_model.py [hunk 16]
```diff
@@ -641,13 +646,413 @@ def get_local_sample_distances(
...
+            "effect_size": effect_size,
+            "pvalue": pvalue,
+            "padj": padj,
+        }
+        if store_lfc:
+            lfc = np.concatenate(lfc, axis=1)
+            lfc = xr.DataArray(
+                lfc,
+                dims=["covariate", "cell_name", "gene"],
+                coords={
```
Comment (justjhong) [inline src/scvi_v2/_model.py]: save coords beforehand since its the same in both arrays

File: src/scvi_v2/_model.py [hunk 17]
```diff
@@ -641,13 +646,413 @@ def get_local_sample_distances(
...
+                lfc.append(np.array(res["lfc"]))
+                baseline_expression.append(np.array(res["baseline_expression"]))
+        beta = np.concatenate(beta, axis=0)
+        effect_size = np.concatenate(effect_size, axis=0)
+        pvalue = np.concatenate(pvalue, axis=0)
+        pvalue_shape = pvalue.shape
+        padj = multipletests(pvalue.flatten(), method="fdr_bh")[1].reshape(pvalue_shape)
+
+        n_latent = beta.shape[2]
+        beta = xr.DataArray(
```
Comment (justjhong) [inline src/scvi_v2/_model.py]: this block of code can be made much more concise by making a dataset in one go, then adding additional data vars in the if cases
see for reference: https://towardsdatascience.com/how-to-create-xarray-datasets-cf1859c95921

**Review by justjhong — COMMENTED** [review]
awesome!

**Review by justjhong — COMMENTED** [review]
awesome!

**Review by justjhong — COMMENTED** [review]
awesome!

File: src/scvi_v2/_module.py (hunk 9 above)
Comment (PierreBoyeau) [reply src/scvi_v2/_module.py]: not at all you're right, i forgot to remove this

File: src/scvi_v2/_model.py (hunk 17 above)
Comment (PierreBoyeau) [reply src/scvi_v2/_model.py]: Wonderful, thanks so much!

File: src/scvi_v2/_model.py (hunk 15 above)
Comment (PierreBoyeau) [reply src/scvi_v2/_model.py:1013]: Yes, I don't know why this works either.

File: src/scvi_v2/_model.py (hunk 4 above)
Comment (PierreBoyeau) [reply src/scvi_v2/_model.py:831]: Would you know how to do this?

File: src/scvi_v2/_model.py (hunk 13 above)
Comment (PierreBoyeau) [reply src/scvi_v2/_model.py:834]: im down but there is one thing i am uncertain about. what if you subsample the original AnnData, such that the `_indices` are not in [0, N-1] anymore?
Where should we make sure that the `_indices` are actually what we neeed?

File: src/scvi_v2/_model.py [hunk 18]
```diff
@@ -641,13 +645,455 @@ def get_local_sample_distances(
...
+            **coord_info,
+        )
+        cell_sample_admissible = log_ratios > admissibility_threshold
+        results = xr.Dataset(
+            {
+                "log_probs": log_probs,
+                "log_ratios": log_ratios,
+                "is_admissible": cell_sample_admissible,
+            }
+        )
```
Comment (PierreBoyeau) [inline src/scvi_v2/_model.py]: Replace by recommended `Dataset` construction (see `perform_multivariate_analysis`)

File: src/scvi_v2/_model.py [hunk 19]
```diff
@@ -641,13 +645,455 @@ def get_local_sample_distances(
...
+                n_components = min(adata_s.n_obs // 4, 20)
+                gmm_ = GaussianMixture(n_components=n_components).fit(adata_s.obsm["U"])
+                log_probs_s = jnp.quantile(gmm_.score_samples(adata_s.obsm["U"]), q=quantile_threshold)
+                log_probs_ = gmm_.score_samples(adata.obsm["U"])[:, None]
+            elif flavor == "ap":
+                ap = self.get_aggregated_posterior(adata=adata_s)
+                log_probs_s = jnp.quantile(ap.log_prob(adata_s.obsm["U"]).sum(axis=1), q=quantile_threshold)
+                log_probs_ = (
+                    self.get_aggregated_posterior(adata=adata).log_prob(adata.obsm["U"]).sum(axis=1, keepdims=True)
+                )
```
Comment (PierreBoyeau) [inline src/scvi_v2/_model.py]: could you make sure this is on CPU, or that it is a numpy object?

File: src/scvi_v2/_model.py [hunk 20]
```diff
@@ -641,13 +645,455 @@ def get_local_sample_distances(
...
+        qu_scale = jnp.concatenate(qu_scales, axis=0).T
+        return dist.MixtureSameFamily(
+            dist.Categorical(probs=jnp.ones(qu_loc.shape[1]) / qu_loc.shape[1]), dist.Normal(qu_loc, qu_scale)
+        )
+
+    def get_outlier_cell_sample_pairs(
+        self,
+        adata=None,
+        flavor: Literal["MoG", "ap"] = "MoG",
+        subsample_size: Optional[int] = None,
```
Comment (PierreBoyeau) [inline src/scvi_v2/_model.py]:
```suggestion
        flavor: Literal["MoG", "ap"] = "ap",
        subsample_size: Optional[int] = None,
```

File: src/scvi_v2/_model.py (hunk 4 above)
Comment (justjhong) [reply src/scvi_v2/_model.py:831]: hm actually that is more annoying than i thought, lets just leave it as is for now

File: src/scvi_v2/_model.py (hunk 13 above)
Comment (justjhong) [reply src/scvi_v2/_model.py:834]: oh thats a good point, for now just making the hack hackier by replacing the indices no matter what so it doesn't get messed up by passing in a subsetted anndata

File: src/scvi_v2/_model.py [hunk 21]
```diff
@@ -739,14 +739,19 @@ def get_outlier_cell_sample_pairs(
...
                 ap = self.get_aggregated_posterior(adata=adata_s)
                 log_probs_s = jnp.quantile(ap.log_prob(adata_s.obsm["U"]).sum(axis=1), q=quantile_threshold)
-                log_probs_ = (
-                    self.get_aggregated_posterior(adata=adata).log_prob(adata.obsm["U"]).sum(axis=1, keepdims=True)
-                )
+                n_splits = adata.n_obs // minibatch_size
+                log_probs_ = []
+                for u_rep in np.array_split(adata.obsm["U"], n_splits):
+                    log_probs_.append(
+                        jax.device_put(ap.log_prob(u_rep).sum(-1, keepdims=True))
```
Comment (justjhong) [inline src/scvi_v2/_model.py]: if we want it on CPU always, this should be device_get right? if default device is set to GPU, this will put it on gpu

**Review by justjhong — COMMENTED** [review]
thanks!

**Review by justjhong — COMMENTED** [review]
thanks!

**Review by justjhong — APPROVED** [review]
Lets goo
